daisy/workflow.go

```diff
@@ -894,7 +887,22 @@ func (w *Workflow) IterateWorkflowSteps(cb func(step *Step)) {
 // CancelWithReason cancels workflow with a specific reason. The specific reason replaces "is canceled" in the default error message.
 func (w *Workflow) CancelWithReason(reason string) {
 	w.cancelReason = reason
-	close(w.Cancel)
+	w.CancelWorkflow()
+}
+
+// CancelWorkflow cancels the workflow. Safe to call multiple times.
+// Prefer this to closing the w.Cancel channel,
+// which will panic if it has already been closed.
+func (w *Workflow) CancelWorkflow() {
+	w.isCanceledMx.Lock()
+	defer w.isCanceledMx.Unlock()
+
+	if !w.isCanceled {
```

Comment:
Why not just keep the isCanceled() function and not use this bool value so you don't need the recover here?

Reply:
AFAICT,  `isCanceled` indicated that either the channel was closed, or a value was written to it. Given the ambiguity, I thought it was clearer to consolidate to a boolean value.